### ISSUE TITLE
Median/percentiles algorithm

### ISSUE THREAD

**Issue by SachaG** [SachaG/tokyodev#79]
Here's a typical dataset ("Scenario A"):

```
3500000
3500000
5500000
7500000
8500000
11500000
14500000
30000000
```

Finding the median is simple enough. Here's a description of the algorithm from Khan Academy:

> Arrange the data points from smallest to largest.
> If the number of data points is odd, the median is the middle data point in the list.
> If the number of data points is even, the median is the average of the two middle data points in the list.

The problem is that the data format we're dealing with is more like this ("Scenario B"):

```
{ value: 3500000, count: 10 }, 
{ value: 5500000, count: 34 }, 
{ value: 7500000, count: 23 }, 
{ value: 8500000, count: 78 }, 
{ value: 11500000, count: 131 }, 
{ value: 14500000, count: 25 }, 
{ value: 30000000, count: 54 }
```

### Option 1

We could expand out the array so that we end up back in Scenario A. The issue is that for larger surveys we'd be dealing with arrays with 10k+ items. That being said, it's probably not that big a deal to handle that for a modern server. 

### Option 2

The other method (which is the current strategy) is to use an accumulator:

```
// use totalCount to simulate what the indexes of the 25th, 50th, and 75th percentiles 
// would be in the "expanded" buckets
const index25 = Math.floor(totalCount * 0.25)
const index50 = Math.floor(totalCount * 0.5)
const index75 = Math.floor(totalCount * 0.75)

for (let i = 0; i < buckets.length; i++) {
        accumulatedCount += buckets[i].count

        if (p25 === null && accumulatedCount >= index25) {
            // as soon as the accumulated count crosses index25, set p25
            p25 = buckets[i].average
        }
        if (p50 === null && accumulatedCount >= index50) {
            p50 = buckets[i].average
        }
        if (p75 === null && accumulatedCount >= index75) {
            p75 = buckets[i].average
        }
        if (p25 !== null && p50 !== null && p75 !== null) {
            break
        }
    }
```

Downside: this makes the logic harder to follow, and maybe risks introducing inaccuracies? Also doesn't currently handle the odd/even bucket length distinction.

**Comment by pwim** [comment]
Finding the percentile of a list that has a large number of items should still take constant time, if you were to sort the buckets and then expand them, but you're right that it would waste a lot of memory.

I was a bit confused your use of `average` in the accumulator approach. For a salary bucket, would that be the midpoint (e.g. `3500000`)?

But I think overall, that the accumulator approach is how I'd do it. Implementation wise, I'd extract a percentile function, and call it three times though. The number of buckets is going to be relatively small, and so it shouldn't make much of a difference if you iterate through the one time or three times. But the code could be a lot cleaner, especially when you're handling the even case.

**Comment by SachaG** [comment]
The reason I need the average is just to sort the buckets. Since they can have names like `range_2000_3000`, I need to first figure out that the average is `2500` before I can actually do anything with them. So yes it would be the midpoint of the bucket, at least in most case.

Since it's an arbitrarily-defined value there are also scenarios where it's not the midpoint. For example if we have a "Over 9000" bucket we'd need to pick `9000` or maybe `10000` as the value that represents the bucket. I don't know if `average` is a good name for that or not though…

**Comment by pwim** [comment]
How are you handling calculation of the percentiles for something like age, where we get a raw number, and then group it into a bucket?

**Comment by SachaG** [comment]
> How are you handling calculation of the percentiles for something like age, where we get a raw number, and then group it into a bucket?

In this case it's easier because the bucket for "people who are 45 years old" would be e.g. `{id: 45, count: 15 }`. So that "average" value that represents the bucket is just the actual id of the bucket.

Grouping is where it gets trickier, if we have the `range_40_50` age group for example we need to take all the buckets that are contained within the group and then calculate percentiles based on those, in effect "ungrouping" the buckets before calculating percentiles.

**Comment by pwim** [comment]
I was wanting to make sure if we did a facet where we showed the median age by prefecture, it would use the original values to calculate the median, and not assume `range_40_50` was 45 for instance. It sounds like that's something you're already considering.

**Comment by SachaG** [comment]
Yes exactly, it did work like that before but I've fixed that since by having each group store the list of the buckets within, and using those for all calculations instead of just the group's rough midpoint.
